core/src/subgraph/registrar.rs

```diff
@@ -401,22 +411,83 @@ fn create_subgraph_version(
+        None => None,
+        Some(_) => panic!("subgraph entity has invalid type in pendingVersion field"),
+    };
     ops.push(EntityOperation::AbortUnless {
-        description: "Subgraph entity must still exist, have same name and currentVersion"
-            .to_owned(),
+        description:
+            "Subgraph entity must still exist, have same name/currentVersion/pendingVersion"
+                .to_owned(),
         query: SubgraphEntity::query().filter(EntityFilter::And(vec![
-            EntityFilter::Equal("name".to_owned(), name.to_string().into()),
-            EntityFilter::Equal(
-                "currentVersion".to_owned(),
-                current_version_id_opt
-                    .clone()
-                    .map(Value::String)
-                    .unwrap_or(Value::Null),
-            ),
+            EntityFilter::new_equal("name", name.to_string()),
+            EntityFilter::new_equal("currentVersion", current_version_id_opt.clone()),
+            EntityFilter::new_equal("pendingVersion", pending_version_id_opt.clone()),
         ])),
         entity_ids: vec![subgraph_entity_id.clone()],
     });
 
+    // Look up current version's deployment hash
+    let current_version_hash_opt = match current_version_id_opt {
+        Some(ref current_version_id) => {
+            let version_entity = store
+                .get(SubgraphVersionEntity::key(current_version_id.clone()))?
+                .ok_or_else(|| {
+                    TransactionAbortError::Other(format!("Subgraph version entity missing"))
+                })
+                .map_err(StoreError::from)?;
+
+            Some(
+                SubgraphDeploymentId::new(
+                    version_entity
+                        .get("deployment")
+                        .unwrap()
+                        .to_owned()
+                        .as_string()
+                        .unwrap(),
+                )
+                .unwrap(),
+            )
+        }
+        None => None,
+    };
+
+    // Look up pending version's deployment hash
+    let pending_version_hash_opt = match pending_version_id_opt {
+        Some(ref pending_version_id) => {
+            let version_entity = store
+                .get(SubgraphVersionEntity::key(pending_version_id.clone()))?
+                .ok_or_else(|| {
+                    TransactionAbortError::Other(format!("Subgraph version entity missing"))
+                })
+                .map_err(StoreError::from)?;
+
+            Some(
+                SubgraphDeploymentId::new(
+                    version_entity
+                        .get("deployment")
+                        .unwrap()
+                        .to_owned()
+                        .as_string()
+                        .unwrap(),
+                )
+                .unwrap(),
+            )
+        }
+        None => None,
+    };
+
+    // Find all subgraph version entities that point to this hash or a hash
```

Comment:
I'm working on a write-up for this, will post in a few minutes

Reply:
So, there are three relevant entity types here: Subgraph, SubgraphVersion, and SubgraphDeploymentAssignment. Subgraph and SubgraphVersion are straightforward to add/update/remove. The challenge is to make sure that SubgraphDeploymentAssignments are created/removed properly.

For the most part, you can think of SubgraphDeploymentAssignments as being derived from Subgraphs and SubgraphVersions. In other words, you could write a SQL query that looks only at Subgraphs and SubgraphVersions and tells you exactly what SubgraphDeploymentAssignments should exist right now. Specifically, a SubgraphDeploymentAssignment should exist for a hash if and only if there exists a SubgraphVersion that points to that hash **and** is either current or pending.

What makes this difficult is that there are many cases where SubgraphVersions can be added/removed, or change from being pending or current to not being pending or current. Before this PR, I wrote separate logic to handle each case:
- When a SubgraphVersion was removed, I would check to see if there are any other versions pointing to that hash, filter them down to the versions that are "current" (or "pending" now with this PR), and if no, I would remove the assignment as well.
- When a SubgraphVersion is created, I would check to see if the previously current version's assignment needs to be removed. Similar check: look for other versions pointing to that hash, filter them down to versions that are "current", exclude the version that is about to become non-current, and if none found, remove the assignment.

This PR is adding more cases, so more logic like this would be needed. It struck me that this was getting redundant. I mentioned earlier the rule for when an assignment should exist: "a SubgraphDeploymentAssignment should exist for a hash if and only if there exists a SubgraphVersion that points to that hash **and** is either current or pending". In each case, I was essentially re-encoding this rule for each set of circumstances, and thinking through the logic for each set of circumstances was getting rather difficult, particularly in edge cases (e.g. a new subgraph version is being created, but it has the same hash as the previous current version -- therefore the assignment shouldn't be removed, and no new one should be created).

So instead, I refactored to add two common methods:
- `read_subgraph_version_summaries`: Reads information from the store about Subgraph and SubgraphVersion entities, produces summary objects.
- `reconcile_assignments`: Looks at two sets of information about Subgraph and SubgraphVersion entities (the summary objects from above) and decides what assignments should be removed or created. The first set is a "before" view (how things are now), and the second set is an "after" view (what the Subgraph and SubgraphVersion entities are being changed to).

The rule for determining whether an assignment should exist is now implemented only in one place: `reconcile_assignments`. The "before" view is easy to get using `read_subgraph_version_summaries`. The difficult part is computing the "after" view. In each place that Subgraph and SubgraphVersion entities are added/updated/removed, these three steps must be followed in order to ensure the right SubgraphDeploymentAssignments exist:
- Call `read_subgraph_version_summaries` to get the "before" state.
- Figure out what the "after" state will look like by applying some changes to the summary objects. I called this "simulating" in the comments.
- Call `reconcile_assignments` with the before and after states.